fv3core/stencils/remapping_part2.py

```diff
@@ -27,125 +25,150 @@ def init_phis(hs: FloatField, delz: FloatField, phis: FloatField, te_2d: FloatFi
-    )
-    zsum1 = utils.make_storage_from_shape(
-        pt.shape[0:2], grid.compute_origin(), cache_key="remapping_part2_zsum1"
-    )
-    if spec.namelist.do_sat_adj:
-        fast_mp_consv = not do_adiabatic_init and consv > constants.CONSV_MIN
+class Remapping_Part2:
+    def __init__(self, pfull):
+        self.grid = spec.grid
+        self.namelist = spec.namelist
+
+        self._copy_from_below_stencil = FrozenStencil(
+            copy_from_below,
+            origin=self.grid.compute_origin(),
+            domain=self.grid.domain_shape_compute(),
+        )
+
+        self._moist_cv_last_step_stencil = FrozenStencil(
+            moist_cv.moist_pt_last_step,
+            origin=(self.grid.is_, self.grid.js, 0),
+            domain=(self.grid.nic, self.grid.njc, self.grid.npz + 1),
+        )
+
+        self._basic_adjust_divide_stencil = FrozenStencil(
+            basic.adjust_divide_stencil,
+            origin=self.grid.compute_origin(),
+            domain=self.grid.domain_shape_compute(),
+        )
+
+        self.do_sat_adjust = self.namelist.do_sat_adj
+
         # TODO pfull is a 1d var
-        kmp = grid.npz - 1
+        self.kmp = self.grid.npz - 1
         for k in range(pfull.shape[0]):
             if pfull[k] > 10.0e2:
-                kmp = k
+                self.kmp = k
                 break
-    if last_step and not do_adiabatic_init:
-        if consv > constants.CONSV_MIN:
-            raise NotImplementedError(
-                "We do not support consv_te > 0.001 "
-                "because that would trigger an allReduce"
-            )
-        elif consv < -constants.CONSV_MIN:
-            raise Exception(
-                "Unimplemented/untested case consv("
-                + str(consv)
-                + ")  < -CONSV_MIN("
-                + str(-constants.CONSV_MIN)
-                + ")"
-            )
 
-    if spec.namelist.do_sat_adj:
-
-        kmp_origin = (grid.is_, grid.js, kmp)
-        kmp_domain = (grid.nic, grid.njc, grid.npz - kmp)
-        saturation_adjustment(
-            te,
-            qvapor,
-            qliquid,
-            qice,
-            qrain,
-            qsnow,
-            qgraupel,
-            qcld,
-            hs,
-            peln,
-            delp,
-            delz,
-            q_con,
-            pt,
-            pkz,
-            cappa,
-            r_vir,
-            mdt,
-            fast_mp_consv,
-            last_step,
-            akap,
-            kmp,
-        )
-        if fast_mp_consv:
-            sum_te(te, te0_2d, origin=kmp_origin, domain=kmp_domain)
-    if last_step:
-        moist_cv.compute_last_step(
-            pt, pkz, dtmp, r_vir, qvapor, qliquid, qice, qrain, qsnow, qgraupel, gz
-        )
-    else:
-        basic.adjust_divide_stencil(
-            pkz, pt, origin=grid.compute_origin(), domain=grid.domain_shape_compute()
+        self._saturation_adjustment = SatAdjust3d()
```

Comment:
you could now pass kmp to the init of saturation adjustment and make that a Frozen stencil as well. can be done in another PR if you prefer.